### PR TITLE
feat(allowlist): add Verilog, SystemVerilog, and VHDL support

### LINKED ISSUE #1049: feat(allowlist): add Verilog, SystemVerilog, and VHDL support
### Problem Statement

Add first-class Hardware Description Language support to OpenCodeReview.

The implementation will:

- Add `.v`, `.sv`, `.vhd`, and `.vhdl` to the supported-file allowlist.
- Exclude conventional testbench files using:
  - `**/tb_*.{v,sv,vhd,vhdl}`
  - `**/*_tb.{v,sv,vhd,vhdl}`
- Add positive, case-insensitive, exclusion, boundary, and negative tests.
- Route `**/*.{v,sv}` to a Verilog/SystemVerilog review-rule document.
- Route `**/*.{vhd,vhdl}` to a VHDL review-rule document.
- Add review guidance covering common HDL correctness, synthesis, concurrency, clock/reset, and simulation-versus-synthesis problems.

From a user's perspective, HDL files included in a change will be recognized and reviewed with language-specific guidance, while conventional testbench files will be excluded.

### Proposed Solution

Add first-class Hardware Description Language support to OpenCodeReview.

The implementation will:

- Add `.v`, `.sv`, `.vhd`, and `.vhdl` to the supported-file allowlist.
- Exclude conventional testbench files using:
  - `**/tb_*.{v,sv,vhd,vhdl}`
  - `**/*_tb.{v,sv,vhd,vhdl}`
- Add positive, case-insensitive, exclusion, boundary, and negative tests.
- Route `**/*.{v,sv}` to a Verilog/SystemVerilog review-rule document.
- Route `**/*.{vhd,vhdl}` to a VHDL review-rule document.
- Add review guidance covering common HDL correctness, synthesis, concurrency, clock/reset, and simulation-versus-synthesis problems.

From a user's perspective, HDL files included in a change will be recognized and reviewed with language-specific guidance, while conventional testbench files will be excluded.

### Alternatives Considered

1. Continue treating HDL files as unsupported. This would cause `.v`, `.sv`, `.vhd`, and `.vhdl` files to be skipped entirely.

2. Add only the extensions without language-specific review rules. This would make the files reviewable, but they would receive only generic review guidance.

3. Perform content-based detection to distinguish Verilog `.v` files from V-language `.v` files. The current rule resolver is extension-based, so content detection would require a significantly larger architectural change and is outside the scope of this contribution.

### Affected Area

Configuration

### Additional Context

This feature is one of the requested language groups in tracking issue #470.

The `.v` extension is shared by Verilog and the V programming language. Following the Hardware Description scope in #470, this proposal treats `.v` as Verilog. The current resolver is extension-based and cannot distinguish the two languages by file content.

Related tracking issue: #470

### PR DESCRIPTION
**Pull request by Parikalp-Bhardwaj**
## Description

Implements the Hardware Description Language group requested in #470, adding support for Verilog, SystemVerilog, and VHDL.

Changes include:

- Allow `.v`, `.sv`, `.vhd`, and `.vhdl` files, including case-insensitive variants.
- Exclude conventional testbench files using:
  - `**/tb_*.{v,sv,vhd,vhdl}`
  - `**/*_tb.{v,sv,vhd,vhdl}`
- Map `.v` and `.sv` files to `verilog.md`.
- Map `.vhd` and `.vhdl` files to `vhdl.md`.
- Add language-specific review guidance for Verilog/SystemVerilog and VHDL.
- Add allowlist, exclusion, negative-boundary, and rule-resolution tests.

### `.v` extension note

The `.v` extension is shared with the V programming language. Following the Hardware Description scope in #470, this change treats `.v` as Verilog. The current resolver uses file extensions and cannot distinguish the two languages by content.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring (no functional changes)
- [ ] Documentation update
- [ ] CI / Build / Tooling

## How Has This Been Tested?

- [x] `make test` passes locally
- [ ] Manual testing (describe below)

Additional verification:

- `go test ./internal/config/allowlist/... ./internal/config/rules/... -count=1`
- `make check`
- `make test`

All commands pass locally.

## Checklist

- [x] My code follows the project's coding style (`go fmt`, `go vet`)
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have updated the documentation accordingly
- [ ] I have signed the CLA

## Related Issues

Closes #1049 
Part of #470

### CODE REVIEW

**Comment by lizhengfeng101** [comment]
@Parikalp-Bhardwaj Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/alibaba/open-code-review?pullRequest=1050) before we can accept your contribution.

**Comment by Parikalp-Bhardwaj** [comment]
Hi @lizhengfeng101, thank you for reviewing my submission. I have completed and signed the Contributor License Agreement.

**Review by wu21-web — CHANGES_REQUESTED** [review]
Please update the documentation. Your desc marked this as completed, but no site documentation updated.
~~Why do agents forget this every time 🤔~~

File: internal/config/rules/rule_docs/vhdl.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+#### Ranges, Indexing, and Resize
+- Vector indexed or sliced outside its declared range, or with the wrong direction (`to` vs. `downto`), causing a runtime error in simulation and wrong bits in hardware
+- `resize` that unintentionally truncates the high bits of a `signed`/`unsigned`, or zero/sign extension that changes the represented value
+- Off-by-one range arithmetic in `(N-1 downto 0)` declarations and loop bounds, and `others => '0'` aggregates applied to a target of unexpected width
+
+#### Inferred Latches and Process Sensitivity
+- A combinational process whose sensitivity list omits signals it reads, so simulation and synthesis disagree; prefer `process(all)` (VHDL-2008) where available
+- A combinational process that does not assign every output on every path (missing `else`, incomplete `case`/`when`, no default assignment), inferring an unintended latch
+- A `case` without `when others`, or a selected assignment that does not cover all input values, leaving outputs undefined for some cases
```
Comment (wu21-web) [inline internal/config/rules/rule_docs/vhdl.md]:
```suggestion
- A branch within a legal, complete `case` statement that fails to assign an output also assigned by the other branches, causing that output to retain its previous value and infer a latch. Do not require `when others` when every value of the selector subtype is explicitly covered; an incomplete VHDL `case` statement is a compile-time error
```

File: internal/config/rules/rule_docs/vhdl.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+#### Clock and Reset Handling
+- A clocked process using a condition other than `rising_edge(clk)`/`falling_edge(clk)` (e.g. `clk'event and clk = '1'` combined with data in the sensitivity list), blurring the intended synchronous behavior
```
Comment (wu21-web) [inline internal/config/rules/rule_docs/vhdl.md]: The original will produce false positives.

```suggestion
- Clocked logic that tests only the clock level (for example, `if clk = '1'`) instead of an edge, or places data-path assignments outside the intended edge condition. Accept both `rising_edge(clk)`/`falling_edge(clk)` and the conventional `clk'event and clk = '1'`/`clk'event and clk = '0'` forms when their semantics are intentional
```

File: internal/config/rules/rule_docs/vhdl.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#### VHDL Review Principles
+> Favor precision over recall: report only defects likely to change synthesized hardware behavior, cause simulation/synthesis mismatch, or introduce timing hazards in the changed RTL. Account for whether the file is synthesizable design code or a simulation-only model before raising findings, and do not report style that a linter or formatter already handles.
+
+#### numeric_std and Type Usage
+- Arithmetic performed on `std_logic_vector` via the deprecated `std_logic_arith`/`std_logic_unsigned` packages instead of `numeric_std` with `signed`/`unsigned`, giving nonportable or ambiguous results
+- Conversions between `integer`, `signed`, `unsigned`, and `std_logic_vector` done implicitly or with the wrong function (`to_integer`, `to_unsigned`, `to_signed`, `resize`), truncating or mis-signing values
+- Mixing signed and unsigned operands, or comparing values of different widths, where `numeric_std` silently resizes or the result is metavalue-dependent
+- Reliance on `'U'`, `'X'`, or `'Z'` propagation in arithmetic, so simulation produces metavalues that synthesis cannot reproduce
```
Comment (wu21-web) [inline internal/config/rules/rule_docs/vhdl.md]: This section misstates VHDL typing and resizing

```suggestion
- Arithmetic implemented with non-standard vendor packages such as `std_logic_arith` or `std_logic_unsigned` instead of `numeric_std`, where package-specific overloads make the result nonportable or ambiguous
- Incorrect explicit conversions between `integer`, `signed`, `unsigned`, and `std_logic_vector`, including a wrong target width passed to `to_unsigned`, `to_signed`, or `resize`, causing truncation or a changed signed interpretation
- Arithmetic or comparisons whose selected overload and result width do not preserve the intended value; distinguish this from direct `signed`/`unsigned` mixing, which normally requires an explicit conversion and is rejected by analysis
- Reliance on `'U'`, `'X'`, or `'Z'` propagation in arithmetic or control flow where synthesized hardware cannot reproduce the simulated metavalue-dependent behavior
```

Comment (wu21-web) [reply internal/config/rules/rule_docs/vhdl.md]: Note: AI generated this suggestion.

File: internal/config/rules/rule_docs/vhdl.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+#### Ranges, Indexing, and Resize
+- Vector indexed or sliced outside its declared range, or with the wrong direction (`to` vs. `downto`), causing a runtime error in simulation and wrong bits in hardware
+- `resize` that unintentionally truncates the high bits of a `signed`/`unsigned`, or zero/sign extension that changes the represented value
```
Comment (wu21-web) [inline internal/config/rules/rule_docs/vhdl.md]: Why?
```suggestion
- `resize` to a smaller width that discards significant information; for `SIGNED`, account for the defined behavior of retaining the original sign bit together with the rightmost part, and for `UNSIGNED`, dropping the leftmost bits
```

File: internal/config/rules/rule_docs/verilog.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+#### Verilog and SystemVerilog Review Principles
+> Favor precision over recall: report only defects likely to change synthesized hardware behavior, cause simulation/synthesis mismatch, or introduce timing hazards in the changed RTL. Account for whether the file is synthesizable design code or a simulation-only model before raising findings, and do not report style that a linter or formatter already handles. This rule covers both Verilog (`.v`) and SystemVerilog (`.sv`).
+
+#### Blocking and Non-Blocking Assignments
+- Non-blocking assignments (`<=`) used for combinational logic, or blocking assignments (`=`) used for sequential logic inside a clocked `always`/`always_ff` block, where the mixed style changes simulated ordering or infers unintended hardware
+- Blocking and non-blocking assignments to the same signal within one `always` block, producing order-dependent, non-deterministic behavior
+- A signal assigned from more than one `always` block, creating multiple drivers on synthesis and a race in simulation
+- Prefer `always_ff`, `always_comb`, and `always_latch` (SystemVerilog) so the tool checks the intended inference instead of a bare `always @(*)`
```
Comment (wu21-web) [inline internal/config/rules/rule_docs/verilog.md]: Verilog procedural assignments are described as nondeterministic too broadly

```suggestion
- Mixed blocking and non-blocking assignments to the same variable within one procedural block when later statements observe a value different from the intended one because non-blocking updates are deferred; do not describe statement ordering within one block as inherently nondeterministic
- A register or variable intended to have one procedural driver being written by multiple processes, especially in violation of `always_ff`, `always_comb`, or `always_latch` single-writer requirements; do not flag intentional multiple drivers on an appropriate net type without showing an actual conflict
- Misuse of `always_ff`, `always_comb`, or `always_latch` that violates their event-control, assignment, or single-writer semantics; do not report a correctly formed bare `always @(*)` solely as a style issue
```

Comment (wu21-web) [reply internal/config/rules/rule_docs/verilog.md]: This suggestion is generated by AI

File: internal/config/rules/rule_docs/vhdl.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+- Signals that must retain state across reset placed on the reset branch, or registers with no defined reset where a known startup state is assumed
+- Gated or derived clocks where a clock enable is intended, and more than one clock driving the same register
+
+#### Clock-Domain Crossings and Resolved Signals
+- A signal generated in one clock domain and sampled in another without a synchronizer (two-flop for single-bit, handshake or asynchronous FIFO for buses), risking metastability
+- Multi-bit buses crossing domains without gray coding or a handshake, so bits arrive skewed
+- Multiple drivers on a resolved signal (`std_logic`) relied on for wired logic, where the resolution function hides an unintended multi-driver conflict; unintended shared drivers on what should be a point-to-point signal
+
+#### Simulation vs. Synthesis and Unsafe Constructs
+- `wait for`, `after` delays, initial signal values, or `assert`/`report` used to drive behavior in code intended for synthesis, where the tool ignores them and hardware disagrees with simulation
```
Comment (wu21-web) [inline internal/config/rules/rule_docs/vhdl.md]:
```suggestion
- Simulation-oriented constructs such as `wait for`, `after` delays, file operations, or reporting side effects whose behavior is required by the design but is unsupported by the declared target synthesis flow. Do not flag initial signal values or assertions solely by syntax when the target tool documents support or safely ignores verification-only statements
```

Comment (wu21-web) [reply internal/config/rules/rule_docs/vhdl.md]: > [!NOTE]
> This suggestion was generated by AI.

File: internal/config/rules/rule_docs/verilog.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+- Gated, derived, or combinationally generated clocks used where a clock enable is intended, and multiple clocks driving the same register
+- Registers with no reset where the design assumes a known power-on state
+
+#### Clock-Domain Crossings and Races
+- A signal sampled in one clock domain that is driven from another without a synchronizer (two-flop for single-bit control, handshake or asynchronous FIFO for buses), risking metastability
+- Multi-bit buses synchronized bit-by-bit, so bits arrive skewed and produce transient invalid values; use gray coding or a handshake
+- Combinational feedback loops, or read-during-write races on inferred memory without a defined collision policy
+
+#### Simulation vs. Synthesis and Unsafe Constructs
+- `initial` blocks, `#delay` timing, `fork`/`join`, or `force`/`release` used in code intended for synthesis, where the tool ignores them and simulation disagrees with hardware
```
Comment (wu21-web) [inline internal/config/rules/rule_docs/verilog.md]: I think synthesis support is tool-dependent

```suggestion
- Simulation-oriented constructs such as `#delay`, `fork`/`join`, `force`/`release`, or an `initial` block whose behavior is required by the design but is unsupported by the declared target synthesis flow. Do not flag initialization solely by syntax when the target FPGA or synthesis tool documents support for it
```

**Review by wu21-web — CHANGES_REQUESTED** [review]
The wiring is complete (including exclude patterns), the documentation has some problems however.

**Review by wu21-web — APPROVED** [review]

**Comment by Parikalp-Bhardwaj** [comment]
Thank you for the detailed review. I’ve updated the `Verilog/SystemVerilog` and `VHDL` rule documentation to address all the points raised, including case completeness, edge detection, typing and resizing semantics, procedural assignments, and tool-dependent synthesis support.

**Comment by lizhengfeng101** [comment]
merge conflicts @Parikalp-Bhardwaj

**Comment by Parikalp-Bhardwaj** [comment]
Hi @lizhengfeng101  I have resolved the merge conflicts, please review it when you have a chance. Thank you

**Comment by lizhengfeng101** [comment]
Checked the branch out and verified behavior rather than only reading the diff — case-insensitive chain, ordering, and the two test packages all pass. Three things:

**1. Test name doesn't match its data.** `subtract` has no `tb` substring (it's `b`+`t`). `rtl/outbound.v` does, and I verified it's still `false`:

```diff
-{"hdl tb substring mid-name not excluded", "rtl/subtract.v", false},
+{"hdl tb substring mid-name not excluded", "rtl/outbound.v", false},
```

**2. `.vh` is missing** (`IsAllowedExt(".vh") == false`). Verilog headers hold the `` `define `` macros and parameter/width constants — exactly what this rule doc targets. No collision, and I verified `**/*.vh` doesn't match `.vhd`/`.vhdl`:

```diff
-    "**/*.{v,sv}": "verilog.md",
+    "**/*.{v,vh,sv}": "verilog.md",
```

Plus `".vh"` in `supported_file_types.json`, two allowlist cases, one resolve case, and the 4 doc tables. Exclude patterns need no change.

**3. The `.v` collision is acknowledged but not mitigated.** Coq/V files that were skipped as `unsupported_ext` now get reviewed *with* RTL guidance. One line in `verilog.md`, continuing the existing "Account for whether the file is synthesizable design code..." sentence, fixes it with no code:

> The `.v` extension is also used by Coq and by the V language; if the file is not Verilog or SystemVerilog, report nothing HDL-specific and fall back to general review principles.

File: internal/config/rules/system_rules.json [hunk 8]
```diff
@@ -42,6 +42,8 @@
     "**/*.zig": "zig.md",
     "**/*.thrift": "thrift.md",
     "**/*.capnp": "capnp.md",
+    "**/*.{v,sv}": "verilog.md",
```
Comment (wu21-web) [inline internal/config/rules/system_rules.json]:
```suggestion
    "**/*.{v,sv,vh}": "verilog.md",
```

**Comment by Parikalp-Bhardwaj** [comment]
Thanks for the helpful review. I’ve addressed all three points:

- Replaced the misleading `subtract.v` test fixture with `outbound.v`.
- Added `.vh` support to the allowlist, rule mapping, tests, and documentation.
- Added guidance for handling `.v` files that contain Coq or V language code.

I also verified the changes with the focused Go tests, `make check`, `make test`, and the documentation typecheck and build. All checks pass locally.

Hopefully it will stay conflict-free this time!

**Comment by lizhengfeng101** [comment]
> Thanks for the helpful review. I’ve addressed all three points:
> 
> * Replaced the misleading `subtract.v` test fixture with `outbound.v`.
> * Added `.vh` support to the allowlist, rule mapping, tests, and documentation.
> * Added guidance for handling `.v` files that contain Coq or V language code.
> 
> I also verified the changes with the focused Go tests, `make check`, `make test`, and the documentation typecheck and build. All checks pass locally.
> 
> Hopefully it will stay conflict-free this time!

merge conflicts ......

**Comment by Parikalp-Bhardwaj** [comment]
I can see that the conflicts were caused by the recently added solidity and vyper entries in main, since they modify the same allowlist and rule-mapping files as this PR 😅

**Comment by wu21-web** [comment]
> [!TIP]
> Moving your insertions away from the end of any map, JSON gets you less merge conflicts.

**Comment by Parikalp-Bhardwaj** [comment]
@wu21-web The conflicts are now resolved and the updated branch has been pushed.

> Tip
> 
> Moving your insertions away from the end of any map, JSON gets you less merge conflicts.

Good point, thanks for the tip. I kept the HDL entries before the existing solidity/vyper entries when resolving the conflicts, so hopefully the same overlap won’t happen again.

**Review by lizhengfeng101 — COMMENTED** [review]
Nit: `pages/src/content/docs/ko/review-rules.md` has the same rule table but wasn't updated — please add the Verilog/VHDL rows there as well.

File: internal/config/allowlist/allowed_ext_test.go [hunk 9]
```diff
@@ -300,6 +310,25 @@ func TestIsExcludedPath(t *testing.T) {
 		{"gen not suffix", "src/gen/util.go", false},
 		{"pb not suffix", "src/pb/client.go", false},
 
+		// Hardware Description Language testbenches (Verilog/SystemVerilog/VHDL).
```
Comment (wu21-web) [inline internal/config/allowlist/allowed_ext_test.go:313]: You missed the new extension @Parikalp-Bhardwaj 

```suggestion
		// Hardware Description Language testbenches (Verilog/SystemVerilog/VH/VHDL).
```

Comment (wu21-web) [reply internal/config/allowlist/allowed_ext_test.go:313]: @Parikalp-Bhardwaj Can you resolve the threads by the way, and it is getting messy.

Comment (wu21-web) [reply internal/config/allowlist/allowed_ext_test.go:313]: @Parikalp-Bhardwaj ?

Comment (Parikalp-Bhardwaj) [reply internal/config/allowlist/allowed_ext_test.go:313]: @wu21-web Done

Comment (wu21-web) [reply internal/config/allowlist/allowed_ext_test.go:313]: @Parikalp-Bhardwaj No you haven't.

File: internal/config/rules/rule_docs/vhdl.md (hunk 1 above)
Comment (wu21-web) [reply internal/config/rules/rule_docs/vhdl.md]: @Parikalp-Bhardwaj What about these thread. Resolve them thanks.

Comment (wu21-web) [reply internal/config/rules/rule_docs/vhdl.md]: And also please put a Co-author-by: there, I wrote so much documentation for you.

**Comment by Parikalp-Bhardwaj** [comment]
Hi @wu21-web anything else?

**Review by wu21-web — APPROVED** [review]
LGTM

File: internal/config/allowlist/allowed_ext_test.go [hunk 10]
```diff
@@ -300,6 +310,25 @@ func TestIsExcludedPath(t *testing.T) {
 		{"gen not suffix", "src/gen/util.go", false},
 		{"pb not suffix", "src/pb/client.go", false},
 
+		// Hardware Description Language testbenches (Verilog/SystemVerilog/VH/VHDL).
```
Comment (lizhengfeng101) [inline internal/config/allowlist/allowed_ext_test.go]: `.vh` is a Verilog header extension and is not covered by the testbench exclude patterns (`**/tb_*.{v,sv,vhd,vhdl}` / `**/*_tb.{v,sv,vhd,vhdl}`), so mentioning "VH" here is misleading.

```suggestion
		// Hardware Description Language testbenches (Verilog/SystemVerilog/VHDL).
```

**Review by lizhengfeng101 — APPROVED** [review]
LGTM
